From f511c1bf0b36bcb2803e81a21304c7529f0652cc Mon Sep 17 00:00:00 2001
From: Michael Auchter <michael.auchter@ni.com>
Date: Mon, 7 Oct 2019 10:19:16 -0500
Subject: [PATCH 09/37] ni: zynqmp: apply overlays based on attached boards

Detect the boards that are attached, and load the appropriate overlays.

Signed-off-by: Michael Auchter <michael.auchter@ni.com>
---
 board/ni/zynqmp/board.c | 68 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 68 insertions(+)

diff --git a/board/ni/zynqmp/board.c b/board/ni/zynqmp/board.c
index 0161840b25c..aafa8ce24f0 100644
--- a/board/ni/zynqmp/board.c
+++ b/board/ni/zynqmp/board.c
@@ -206,8 +206,70 @@ unsigned long do_go_exec(ulong (*entry)(int, char * const []), int argc,
 	return ret;
 }
 
+/*
+ * dtbo is a format string containing %s (slot) and %u (rev) in that order
+ */
+static const struct {
+	uint16_t pid;
+	const char *dtbo;
+} dtbo_map[] = {
+	{
+		.pid = 0x4001,
+		.dtbo = "ni_x4xx-%s-rfchar-rev%u.dtbo",
+	},
+	{
+		.pid = 0x4003,
+		.dtbo = "ni_x4xx-%s-hdmise-rev%u.dtbo",
+	},
+	{
+		.pid = 0x4004,
+		.dtbo = "ni_x4xx-%s-gpsdo-rev%u.dtbo",
+	},
+	{
+		.pid = 0x4005,
+		.dtbo = "ni_x4xx-%s-nogpsdo-rev%u.dtbo",
+	},
+};
+
+static const char *find_dtbo(struct eeprom_ctx *ctx, uint8_t *rev)
+{
+	const struct usrp_eeprom_board_info *info;
+
+	if (!ctx->valid)
+		return NULL;
+
+	info = tlv_lookup(ctx->eeprom.tlv, ctx->eeprom.size,
+			  USRP_EEPROM_BOARD_INFO_TAG);
+	if (!info)
+		return NULL;
+
+	*rev = info->compat_rev;
+
+	for (size_t i = 0; i < ARRAY_SIZE(dtbo_map); i++)
+		if (info->pid == dtbo_map[i].pid)
+			return dtbo_map[i].dtbo;
+
+	return NULL;
+}
+
+static void append_overlay(char *overlays, struct eeprom_ctx *ctx, const char *slot)
+{
+	const char *dtbo;
+	char overlay[128];
+	uint8_t rev;
+
+	dtbo = find_dtbo(ctx, &rev);
+	if (!dtbo)
+		return;
+
+	sprintf(overlay, dtbo, slot, rev);
+	printf("applying overlay: %s\n", overlay);
+	sprintf(overlays, "%s#conf@%s", overlays, overlay);
+}
+
 int board_late_init(void)
 {
+	char overlays[256] = "";
 	u32 reg = 0;
 	u8 bootmode;
 	int ret;
@@ -246,5 +308,11 @@ int board_late_init(void)
 	load_eeprom("ni,dioaux-eeprom", &dioaux_eeprom);
 	load_eeprom("ni,clkaux-eeprom", &clkaux_eeprom);
 
+	append_overlay(overlays, &db0_eeprom, "db0");
+	append_overlay(overlays, &db1_eeprom, "db1");
+	append_overlay(overlays, &dioaux_eeprom, "dioaux");
+	append_overlay(overlays, &clkaux_eeprom, "clkaux");
+	env_set("overlays", overlays);
+
 	return 0;
 }
-- 
2.43.0

